examples/c/uprobe.c

```diff
@@ -12,14 +12,19 @@ static int libbpf_print_fn(enum libbpf_print_level level, const char *format, va
 	return vfprintf(stderr, format, args);
 }
 
-/* It's a global function to make sure compiler doesn't inline it. */
-int uprobed_add(int a, int b)
+/* It's a global function to make sure compiler doesn't inline it. To be extra
+ * sure we also use "asm volatile" and noinline attributes to prevent
+ * compiler from local inlining.
```

Comment:
What do you mean by local inlining?

Reply:
compiler can choose to inline calls to uprobe_add() function from other functions defined in the same .c file, but still keep uprobe_add() around (because it's global, so could be, technically, called from other .c/.o files). This is probably what's happening here, as uprobe_add seems to be attachable (we get no error finding its offset), but it's not really called at runtime, because all the calls are actually inlined.

For global functions, adding `__attribute__(weak)` should help and be enough. In general, for static functions it's way more problematic. You need `__attribute__((noinline))` and `asm volatile ();` trick (I don't think just one of those is enough). So in this case I think just having weak would be ok, but noinline+asm volatile is kind of universal "big hammer" approach I went with.

Reply:
oh, in case you haven't seen, this was in response to https://github.com/libbpf/libbpf-bootstrap/issues/243